great_expectations/experimental/metric_repository/metrics.py

```diff
@@ -0,0 +1,140 @@
+    """Abstract computed metric. Domain, value and parameters are metric dependent.
+
+    Note: This implementation does not currently take into account
+    other domain modifiers, e.g. row_condition, condition_parser, ignore_row_if
+    """
+
+    def __new__(cls, *args, **kwargs):
+        if cls is Metric:
+            raise NotImplementedError("Metric is an abstract class.")
+        instance = super().__new__(cls)
+        return instance
+
+    id: uuid.UUID = Field(description="Metric id")
+    batch: Batch = Field(description="Batch")
+    metric_name: str = Field(description="Metric name")
+    exception: MetricException = Field(description="Exception info if thrown")
+
+    @classmethod
+    def update_forward_refs(cls):
+        from great_expectations.datasource.fluent.interfaces import Batch
+
+        super().update_forward_refs(
+            Batch=Batch,
+        )
+
+
+# Metric domain types
+
+
+class TableMetric(Metric):
+    def __new__(cls, *args, **kwargs):
+        if cls is TableMetric:
+            raise NotImplementedError("TableMetric is an abstract class.")
+        instance = super().__new__(cls)
+        return instance
+
+
+class ColumnMetric(Metric):
+    def __new__(cls, *args, **kwargs):
+        if cls is ColumnMetric:
+            raise NotImplementedError("ColumnMetric is an abstract class.")
+        instance = super().__new__(cls)
+        return instance
+
+    column: str = Field(description="Column name")
+
+
+# TODO: Add ColumnPairMetric, MultiColumnMetric
+
+
+# Metric value types
+
+
+class NumericMetric(Metric):
+    def __new__(cls, *args, **kwargs):
+        if cls is NumericMetric:
+            raise NotImplementedError("NumericMetric is an abstract class.")
+        instance = super().__new__(cls)
+        return instance
+
+    value: Union[int, float] = Field(description="Metric value")
+
+
+class NumericListMetric(Metric):
+    def __new__(cls, *args, **kwargs):
+        if cls is NumericListMetric:
+            raise NotImplementedError("NumericListMetric is an abstract class.")
+        instance = super().__new__(cls)
+        return instance
+
+    value: List[float] = Field(description="Metric value")
+
+
+class StringListMetric(Metric):
+    def __new__(cls, *args, **kwargs):
+        if cls is StringListMetric:
+            raise NotImplementedError("StringListMetric is an abstract class.")
+        instance = super().__new__(cls)
+        return instance
+
+    value: List[str] = Field(description="Metric value")
+
+
+# Concrete Metrics
+# This is where the concrete metric types are defined that
+# bring together a domain type, value type and any parameters (aka metric_value_kwargs)
+
+# TODO: Add metrics here for all Column Descriptive Metrics
+
+
+class NumericTableMetric(TableMetric, NumericMetric):
+    pass
+
+
+class StringListTableMetric(TableMetric, StringListMetric):
+    pass
+
+
+class QuantileValuesColumnMetric(ColumnMetric, NumericListMetric):
```

Comment:
This is a good example of a Metric that inherits + has its own fields 
* ColumnMetric = the domain 
* NumericListMetric = the quantile values (the actual numbers)
* quantiles, allow_relative_error = Metric-specific kwargs. quantiles here are the params (0, 0.25, 0.5, 0.75 etc).